### DISCUSSION TITLE
Monolithic Serverpod Stress Testing

### DISCUSSION THREAD

**Discussion by tricaricom** [serverpod/serverpod#2630] [Q&A]
Does anyone know how many concurrent users a monolithic Serverpod can handle?  I am building a couple of apps and there is a very real possibility any of these apps could possibly have 1m+ users.

Have any metrics been gathered on each cloud provider (AWS, GCP, Azure) can handle the most concurrent users in a single monolithic Serverpod?

Thanks!

**Comment by Isakdl** [comment]
@abdelaziz-mahdy did a performance benchmark a while back, here: https://github.com/serverpod/serverpod/discussions/1791

But I also think that when it comes to questions like this, it is very dependent on what application you are building, how much work you are expected to perform per user, and how active these users are. 

Some extreme examples would be a tax agency, the number of users are all working people in the country but you are only expecting them to visit your application once a year (filing taxes), vs a real-time trading platform where you are expecting each user to perform queries every second or even faster. The load will be very different even if you have the same number of users.

**Reply by tricaricom** [reply]
Excellent thoughts on this, and yes it is extremely dependent on how the users will use the platform.   My current dilemma is regarding streaming real-time chat.  I could potentially have 100,000s of real time streaming chats going on concurrently around the world.  So, in this case, would a real time streaming option be good, or a timer-based polling to check an API endpoint to get the latest chat message.

**Reply by Isakdl** [reply]
Interesting question, my first thought would be around limitations in the OS, but it seems you are not limited by the hard limit of 64k ports at least, see: https://superuser.com/questions/251596/is-there-a-hard-limit-of-65536-open-tcp-connections-per-ip-address-on-linux/251597#251597

But there will be a limit on the file descriptors, this may perhaps be configurable. 

How this scales in Dart and Serverpod is a good question, I haven't pushed the framework to these limits myself before, hence I cannot give you a concrete answer right now.

**Reply by tricaricom** [reply]
Thanks for your response!  At this time, because of all the uncertainty, I think a timer that checks an endpoint to get new chat data is probably the best option for now.  This was my default plan anyways and I can coding my app for this.
